src/profile-logic/import/instruments/index.js

```diff
@@ -0,0 +1,899 @@
+    if (stackTrace) {
+      for (let index = 0; index < stackTrace.length; index++) {
+        const frameAddress = stackTrace[index];
+        const keyOfStackKeyToIndexMap =
+          typeof parentIndex === 'number'
+            ? '$' + parentIndex + '$' + frameAddress
+            : '$' + frameAddress;
+
+        if (!stackKeyToIndex.has(keyOfStackKeyToIndexMap)) {
+          const frameIndex = frameKeyToIndex.get(frameAddress + '');
+
+          if (
+            typeof frameIndex === 'number' &&
+            typeof parentIndex === 'number'
+          ) {
+            createStack(
+              stackTable,
+              stringTable,
+              stackKeyToIndex,
+              keyOfStackKeyToIndexMap,
+              parentIndex,
+              frameIndex
+            );
+          }
+        }
+
+        parentIndex = stackKeyToIndex.get(keyOfStackKeyToIndexMap);
+
+        if (index === stackTrace.length - 1) {
+          const stackForSample = stackKeyToIndex.get(keyOfStackKeyToIndexMap);
+          if (stackForSample) {
+            samplesTable.stack.push(stackForSample);
+          } else {
+            samplesTable.stack.push(null);
+          }
+          samplesTable.time.push(sample.timestamp / 1000000);
+          samplesTable.responsiveness.push(null);
+          samplesTable.length++;
+        }
+      }
+    }
+  }
+
+  return thread;
+}
+
+// This function creates a thread for each group of samples(by threadID)
+function pushThreadsInProfile(profile, addressToFrameMap, samples) {
+  const threadIDToSamples = new Map();
+  for (const sample of samples) {
+    const samplesArray = threadIDToSamples.get(sample.threadID);
+    if (samplesArray) {
+      samplesArray.push(sample);
+    } else {
+      threadIDToSamples.set(sample.threadID, [sample]);
+    }
+  }
+
+  for (const threadID of threadIDToSamples.keys()) {
+    const samples = threadIDToSamples.get(threadID);
+    let processedThread = getEmptyThread();
+    if (samples) {
+      processedThread = getProcessedThread(
+        threadID,
+        samples,
+        addressToFrameMap
+      );
+    }
+    profile.threads.push(processedThread);
+  }
+}
+
+export async function convertInstrumentsProfile(
+  entry: { name: string },
+  fileReaderHelper: fileReader
+): Promise<Profile> {
+  fileReader = fileReaderHelper;
+  const tree = await extractDirectoryTree(entry);
+  const { runs, instrument } = await readFormTemplateFile(tree);
+
+  if (instrument !== 'com.apple.xray.instrument-type.coresampler2') {
+    throw new Error(
+      `The only supported instrument from .trace import is "com.apple.xray.instrument-type.coresampler2". Got ${instrument}`
+    );
+  }
+
+  const profile = getEmptyProfile();
+
+  const { addressToFrameMap, number } = runs[0];
+  // For now, we will just process the first run
+
+  const samples = await getSamples({
+    tree,
+    addressToFrameMap,
+    runNumber: number,
+  });
+
+  pushThreadsInProfile(profile, addressToFrameMap, samples);
```

Comment:
In this function, we are getting an empty profile and then filling it. Instead we can do this function a pure function by moving the profile creation(the `const profile = getEmptyProfile();` line) inside this function and make that function return that profile instead. We'll need to change the name of the function as well.